File: java-app/src/main/java/com/example/javaapp/model/Task.java
package com.example.javaapp.model;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.UUID;

public class Task {
    private final String id;
    private String title;
    private String description;
    private Category category;
    private TaskStatus status;
    private final String assignedUserId;
    private final LocalDateTime createdDate;
    private LocalDateTime completedDate;

    public Task(String title, String description, Category category, String assignedUserId) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Title cannot be empty");
        }
        if (category == null) {
            throw new IllegalArgumentException("Category cannot be null");
        }
        if (assignedUserId == null || assignedUserId.trim().isEmpty()) {
            throw new IllegalArgumentException("Assigned user ID cannot be empty");
        }

        this.id = UUID.randomUUID().toString();
        this.title = title;
        this.description = description != null ? description : "";
        this.category = category;
        this.status = TaskStatus.PENDING;
        this.assignedUserId = assignedUserId;
        this.createdDate = LocalDateTime.now();
        this.completedDate = null;
    }

    // Getters
    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public Category getCategory() {
        return category;
    }

    public TaskStatus getStatus() {
        return status;
    }

    public String getAssignedUserId() {
        return assignedUserId;
    }

    public LocalDateTime getCreatedDate() {
        return createdDate;
    }

    public LocalDateTime getCompletedDate() {
        return completedDate;
    }

    public String getFormattedCreatedDate() {
        return createdDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
    }

    // Setters with validation
    public void setTitle(String title) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Title cannot be empty");
        }
        this.title = title;
    }

    public void setDescription(String description) {
        this.description = description != null ? description : "";
    }

    public void setCategory(Category category) {
        if (category == null) {
            throw new IllegalArgumentException("Category cannot be null");
        }
        this.category = category;
    }

    public boolean setStatus(TaskStatus newStatus) {
        if (newStatus == null) {
            throw new IllegalArgumentException("Status cannot be null");
        }

        if (status.canTransitionTo(newStatus)) {
            this.status = newStatus;
            if (newStatus == TaskStatus.COMPLETED) {
                this.completedDate = LocalDateTime.now();
            }
            return true;
        }
        return false;
    }

    public boolean isCompleted() {
        return status == TaskStatus.COMPLETED;
    }

    @Override
    public String toString() {
        return String.format("Task{id='%s', title='%s', category=%s, status=%s, assignedTo='%s'}",
                id.substring(0, 8) + "...", title, category, status, assignedUserId.substring(0, 8) + "...");
    }
}

File: js-app/models/Task.js
import { v4 as uuidv4 } from 'uuid';
import { Category, getCategoryByName } from './Category.js';
import { TaskStatus, getStatusByName, canTransitionTo } from './TaskStatus.js';

class Task {
    constructor(title, description, category, assignedUserId) {
        // Validation
        if (!title || title.trim() === '') {
            throw new Error('Title cannot be empty');
        }
        if (!category) {
            throw new Error('Category cannot be null');
        }
        if (!assignedUserId || assignedUserId.trim() === '') {
            throw new Error('Assigned user ID cannot be empty');
        }

        this.id = uuidv4();
        this.title = title;
        this.description = description || '';
        this.category = typeof category === 'string' ? getCategoryByName(category) : category;
        this.status = TaskStatus.PENDING;
        this.assignedUserId = assignedUserId;
        this.createdDate = new Date().toISOString();
        this.completedDate = null;
    }

    // Status update with transition validation
    setStatus(newStatus) {
        if (!newStatus) {
            throw new Error('Status cannot be null');
        }

        const statusObj = typeof newStatus === 'string' ? getStatusByName(newStatus) : newStatus;

        if (canTransitionTo(this.status, statusObj)) {
            this.status = statusObj;
            if (statusObj === TaskStatus.COMPLETED) {
                this.completedDate = new Date().toISOString();
            }
            return true;
        }
        return false;
    }

    isCompleted() {
        return this.status === TaskStatus.COMPLETED;
    }

    // Getters
    getCategoryDisplayName() {
        return this.category.displayName;
    }

    getStatusDisplayName() {
        return this.status.displayName;
    }

    getFormattedCreatedDate() {
        const date = new Date(this.createdDate);
        return date.toLocaleDateString() + ' ' + date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
    }

    // Setters with validation
    setTitle(title) {
        if (!title || title.trim() === '') {
            throw new Error('Title cannot be empty');
        }
        this.title = title;
    }

    setDescription(description) {
        this.description = description || '';
    }

    setCategory(category) {
        if (!category) {
            throw new Error('Category cannot be null');
        }
        this.category = typeof category === 'string' ? getCategoryByName(category) : category;
    }

    setAssignedUserId(assignedUserId) {
        this.assignedUserId = assignedUserId;
    }

    // For JSON serialization
    toJSON() {
        return {
            id: this.id,
            title: this.title,
            description: this.description,
            category: this.category.name,
            status: this.status.name,
            assignedUserId: this.assignedUserId,
            createdDate: this.createdDate,
            completedDate: this.completedDate
        };
    }

    // Create from stored data
    static fromJSON(data) {
        const task = new Task(
            data.title,
            data.description,
            getCategoryByName(data.category),
            data.assignedUserId
        );
        task.id = data.id;
        task.status = getStatusByName(data.status);
        task.createdDate = data.createdDate;
        task.completedDate = data.completedDate;
        return task;
    }

    toString() {
        return `Task{id='${this.id.substring(0, 8)}...', title='${this.title}', category=${this.category.name}, status=${this.status.name}}`;
    }
}

export default Task;